### ISSUE TITLE
🐛 Bug Report: Errors are not logged

### ISSUE THREAD

**Issue by nirga** [traceloop/openllmetry#412]
### Which component is this bug for?

All Packages

### 📜 Description

If an HTTP error is returned from a foundation model API, we don't properly log it as a failed span.

### 👟 Reproduction steps

N/A

### 👍 Expected behavior

N/A

### 👎 Actual Behavior with Screenshots

N/A

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### Are you willing to submit PR?

None

**Comment by HappyLife2** [comment]
Hi @nirga. Could you specify which components or packages within the project interact directly with foundation model APIs? This information will help narrow down the search to the relevant parts of the codebase. 

Also, could you provide insight into how a 'failed span' is defined or identified within the context of the project? This will aid in accurately logging the relevant events as failed spans.

**Comment by nirga** [comment]
Hey @HappyLife2!
These are `opentelemetry-instrumentatin-openai`, `-anthropic`, `-cohere`, `-replicate`, `-bedrock`, `-vertexai` and `-transformers`.

This is how a failed span should be logged: https://opentelemetry.io/docs/languages/python/instrumentation/#record-exceptions-in-spans

**Comment by epg323** [comment]
I can be assigned to this one if no one wants it.

**Comment by nirga** [comment]
That would be great @epg323 :) 
I assume @HappyLife2 is not actively working on this since it's been 2 months.

**Comment by nirga** [comment]
Yes @akhilk2802! Appreciate any help :)

**Comment by akhilk2802** [comment]
okay, looking into it

**Comment by Huzaifa666** [comment]
Is this issue still open or resolved ?

**Comment by nirga** [comment]
Yes this is still open, not sure what's the status of @akhilk2802 though

**Comment by Huzaifa666** [comment]
Hey @nirga 

I was able to setup a sample project and can see the spans in my dashboard
![Screenshot from 2024-07-20 23-20-42](https://github.com/user-attachments/assets/11944379-56fc-4fda-a868-59e576fb0989)

I produced a 401 error code by inputting incorrect API_KEY 

![Screenshot from 2024-07-20 23-23-54](https://github.com/user-attachments/assets/e034125b-cc08-40ab-b6f6-219998085f39)

You want this error to be logged as a span
Correct me if I am wrong here

**Comment by nirga** [comment]
Exactly @Huzaifa666!

**Comment by Huzaifa666** [comment]
![Screenshot from 2024-07-20 23-51-53](https://github.com/user-attachments/assets/7e5734a9-cf15-4612-ac7d-2730577af9a3)

Is this acceptable ?
@nirga

**Comment by nirga** [comment]
Looks good, let's look at the code

**Comment by Huzaifa666** [comment]
Have linked the PR with the issue
Shared in the slack channel as well

**Comment by Bommy32** [comment]
Is this still open by chance? I'd love to give it a go. 

If it is, I would like to clarify please: Is it for all packages everywhere or just all packages for the ones referenced earlier? opentelemetry-instrumentatin-openai, -anthropic, -cohere, -replicate, -bedrock, -vertexai and -transformers

**Comment by nirga** [comment]
@Bommy32 yes it's still open :) 
It's for all packages, except for the SDK and the sample app

**Comment by Bommy32** [comment]
@nirga Wonderful! Thank you.

Just a heads up as well - I am still rather new and learning so I may end up being entirely out to lunch in some aspects but I am going to try my best and make any adjustments needed so I can truly learn.

**Comment by jareks** [comment]
Any progress on this please? Logging errors seems like very basic functionality to have.

**Comment by nirga** [comment]
@jareks any specific pacakges / instrumentation you'd like us to prioritize?

**Comment by jareks** [comment]
Exceptions from openai client would be great 👍

**Comment by apthagowda97** [comment]
Any update on langchain?

**Comment by nirga** [comment]
@apthagowda97 we're adding this now! Sorry for the delay here

**Comment by shristirwt** [comment]
Hi @nirga , 
This issue is still open and I would like to help. I see that PRs for openai and langchain have already been merged.
I can take on one of the remaining packages. Would it be helpful if I started with opentelemetry-instrumentation-cohere (or pick another one from the list)? 
Please let me know!

**Comment by nirga** [comment]
Yes that would be perfect @shristirwt

**Comment by shristirwt** [comment]
Hi @nirga ,
I've finished the changes locally to fix the error logging for the cohere package. I've added try...except...finally blocks to the streaming functions and fixed the async wrapper. I also added new unit tests for the error paths, and all tests for the package are passing.

Since this PR only fixes the issue for cohere (and the original issue #412 covers multiple packages), should I open the Pull Request linked to #412, or would you prefer I create a new, separate issue specifically for the cohere fix and link the PR to that?

Let me know!

**Comment by nirga** [comment]
Small PRs are always better :) @shristirwt

**Comment by DDiyash** [comment]
Hey @nirga, I see that the issue is still open, I’d like to take on opentelemetry-instrumentation-anthropic. Could you assign this issue to me?

**Comment by srijansingh9170-source** [comment]
Hi, I would like to work on this issue. Could you please assign it to me?
